six/modules/c++/six.sicd/source/Grid.cpp

```diff
@@ -84,4 +471,831 @@ bool Grid::operator==(const Grid& rhs) const
+            << "Grid.Row.UVectECF: " << row->unitVector << std::endl
+            << "Derived Grid.Row.UVectECF: "
+            << derivedRowUnitVector(rmcr, scp);
+        log.error(messageBuilder.str());
+        valid = false;
+    }
+
+    // 2.12.3.3.4
+    if ((col->unitVector - derivedColUnitVector(rmcr, scp)).norm() > UVECT_TOL)
+    {
+        messageBuilder.str("");
+        messageBuilder << "UVect fields inconsistent." << std::endl
+            << "Grid.Col.UVectECF: " << col->unitVector << std::endl
+            << "Derived Grid.Col.UVectECF: "
+            << derivedRowUnitVector(rmcr, scp);
+        log.error(messageBuilder.str());
+        valid = false;
+    }
+
+    // 2.12.3.3.6
+    if (col->kCenter != 0)
+    {
+        messageBuilder.str("");
+        messageBuilder << "Grid.Col.KCtr must be zero for RMA/RMCR data." 
+            << std::endl << "Grid.Col.KCtr = " << col->kCenter;
+        log.error(messageBuilder.str());
+        valid = false;
+    }
+
+    // 2.12.3.3.7
+    if (!Init::isUndefined<double>(fc))
+    {
+        if (std::abs(row->kCenter / derivedRowKCenter(rmcr, fc) - 1) > WF_TOL)
+        {
+            messageBuilder.str("");
+            messageBuilder << WF_INCONSISTENT_STR << std::endl
+                << "Grid.Row.KCtr: " << row->kCenter << std::endl
+                << "Center frequency * 2/c: " << derivedRowKCenter(rmcr, fc);
+            log.warn(messageBuilder.str());
+            valid = false;
+        }
+    }
+
+    return valid;
+}
+
+double Grid::derivedRowKCenter(const RMCR& rmcr, double fc) const
+{
+    return fc * (2 / math::Constants::SPEED_OF_LIGHT_METERS_PER_SEC);
+}
+
+bool Grid::validate(const INCA& inca, const Vector3& scp,
+        const PolyXYZ& arpPoly, double fc,
+        logging::Logger& log) const
+{
+    (void)inca; // Only for overloading
+    bool valid = true;
+    std::ostringstream messageBuilder;
+    const double IFP_POLY_TOL = 1e-5;
+
+    if (!Init::isUndefined<Poly2D>(inca.dopplerCentroidPoly) &&
+        inca.dopplerCentroidCOA == BooleanType::IS_TRUE)
+    {
+        const Poly2D& kcoaPoly = col->deltaKCOAPoly;
+        const Poly2D& centroidPoly = inca.dopplerCentroidPoly;
+
+        if (kcoaPoly.orderX() != centroidPoly.orderX() &&
+            kcoaPoly.orderY() != centroidPoly.orderY())
+        {
+            valid = false;
+            messageBuilder.str("");
+            messageBuilder << "Grid.Col.delaKCOAPoly and "
+                << "RMA.INCA.dopplerCentroidPoly have diferent sizes.";
+            log.error(messageBuilder.str());
+        }
+        else
+        {
+            Poly2D differencePoly = kcoaPoly - (centroidPoly * inca.timeCAPoly[1]);
+            double norm = 0;
+
+            for (size_t ii = 0; ii < differencePoly.orderX(); ++ii)
+            {
+                for (size_t jj = 0; jj < differencePoly.orderY(); ++jj)
+                {
+                    norm += std::pow(differencePoly[ii][jj], 2);
```

Comment:
Here and all the other spots that you use `std::pow(someValue, 2)`... `std::pow()` is generally dumbly inefficient in cases like this.  It is a lot more efficient to have...

```
inline double square(double val)
{
    return val * val;
}
```

In fact, if you grep through the SIX repo, you'll see that we've done that in a few cpp files.  Go ahead and add this function into the `math` module too and then replace all the uses of it in SIX as well as the `std::pow(someValue, 2)` cases you've added with it.

Not that efficiency really matters here, but it's no more complex to use this instead so we might as well.